Refactor the constituency lookup component

asibs opened the pull request:
* Allow concurrent calls to lookupPostcode:
* Calls are throttled (typing lots of postcodes will just lookup the last one)
* Results are cached

* Put a spinner on the button whilst requests are in progress
* Display the currently chosen constituency.
* Gray out postcodes which are invalid in the box.

jms301 committed:
Refactor the constituency lookup component

* Allow concurrent calls to lookupPostcode:
* Calls are throttled (typing lots of postcodes will just lookup the last one)
* Results are cached

* Put a spinner on the button whilst requests are in progress
* Display the currently chosen constituency.
* Gray out postcodes which are invalid in the box.

asibs commented on components/constituency_lookup/ConstituencyLookup.tsx:
```diff
@@ -25,22 +34,93 @@ const errorCodeToErrorMessage = (code: ErrorCode) => {
+const throttledApi = async (
+  requestBody: ConstituencyLookupRequest,
+): Promise<ConstituencyLookupResponse | null> => {
+  if (lookupCache.hasOwnProperty(requestBody.postcode)) {
```
This doesn't matter at the moment, but if/when we're on the DemoClub API for uncertain postcodes, then we'll need to check if the `ConstituencyLookupRequest` includes an `addressSlug` and if not, don't use the cache (or we make the cache a two-layer thing: postcode -> address (can be null) -> results.

Can just add a TODO to handle this if/when we start using the DC API.

jms301 replied:
Yeah good point.

asibs commented on components/constituency_lookup/ConstituencyLookup.tsx:
```diff
@@ -25,22 +34,93 @@ const errorCodeToErrorMessage = (code: ErrorCode) => {
+): Promise<ConstituencyLookupResponse | null> => {
+  if (lookupCache.hasOwnProperty(requestBody.postcode)) {
+    const cached = await lookupCache[requestBody.postcode];
+    if (cached !== null) {
```
Can just use `if (cached) {` and this should deal with nulls, undefined, etc?

asibs commented on components/constituency_lookup/ConstituencyLookup.tsx:
```diff
@@ -49,123 +129,111 @@ const PostcodeLookup = () => {
+  const [lastValidPcode, setLastValidPcode] = useState<string>("");
+  const [constituencyIdx, setConstituencyIdx] = useState<number | false>(false);
+
+  let userConstituency: Constituency | null = null;
+
+  if (
+    apiResponse &&
+    apiResponse.constituencies &&
+    apiResponse.constituencies.length != 0 &&
+    apiResponse.postcode == lastValidPcode &&
+    constituencyIdx !== false
+  ) {
+    userConstituency = apiResponse.constituencies[constituencyIdx];
+  }
```
I _think_ we could use `useMemo` for this with `apiResponse` & `constituencyIdx` as the dependencies, unless that also suffers from the same timing issues you mentioned?

jms301 replied:
So from how I understand the problem useMemo also won't work. 

I think the problem is basically fetch calls can still be waiting to return inside the scope of a ConstituencyLookup component that's now got a stale postcode. So any state they try to check their response against will just be the state that existed when they started? 

Even if there was some way to detect that the component which called the fetch is now stale or even if the postcode is stale it could've switched away and back. 

Probably the 'correct' solution is to hold that state in the parent. But I think this is good enough as I can't really imagine having two of these on the same page.

asibs commented on components/constituency_lookup/ConstituencyLookup.tsx:
```diff
@@ -25,22 +34,93 @@ const errorCodeToErrorMessage = (code: ErrorCode) => {
+    reqControl.time = Date.now();
+    lookupCache[requestBody.postcode] = fetchApi(requestBody);
+  } else {
+    //need to delay the request
```
Doesn't this end up meaning we still hit the API the same number of times (minus caching), just we defer multiple API calls which otherwise would've happened in quick succession?

An alternative might be that if the rate-limit has been hit, don't make the API call at all, and wait for the user to click the submit button before doing the lookup?

jms301 replied:
The call to clearTimeout means you cancel previous requests that were issued within rateLimit time. So at most you can issue 1 request per rateLimit time period.

asibs replied:
Oh yeah, sorry, missed that line above :+1:

asibs reviewed (commented):
Haven't done a full review, but added some comments for consideration

jms301 commented on components/constituency_lookup/ConstituencyLookup.tsx:
```diff
@@ -25,22 +34,93 @@ const errorCodeToErrorMessage = (code: ErrorCode) => {
+
+    if (response.ok) {
+      lookupCache[requestBody.postcode] = response.json();
+      return lookupCache[requestBody.postcode];
```
This should just return, the parent function is dealing with the Cache.

jms301 committed:
Switch constituency_lookup from POST to GET request.

Also fix misc comments & minor code issues
w

asibs commented on app/api/constituency_lookup/[postcode]/[[...address]]/route.ts:
```diff
@@ -18,27 +18,25 @@ console.log(`***** dbPath is [${dbPath}] *****`);
 // constituency_lookup/{postcode}/{addressSlug}
 // which should enable caching
-export async function POST(request: NextRequest) {
+export async function GET(request: NextRequest, context) {
```
It seems like if you use the `request` param, the route will be opted-out of caching :thinking: 
https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

The dynamic route segments section doesn't say anything about caching:
https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

It may be that if we have `request` in the function signature, but don't access it, caching will still work, but not sure.
We're not actively using it anymore tho, so we can remove the `console.log(request)` a few lines down - I suspect having that will _definitely_ stop caching, but maybe the dynamic routes also prevents caching :roll_eyes:

asibs replied:
Different topic on this line - the Vercel deploy is failing with:
`Type error: Parameter 'context' implicitly has an 'any' type.`

I probably need to add a typescript check to the linter/pretty Github action check.

asibs committed:
Remove some logging

jms301 committed:
mark address param as optional, make it clearer where params are used

asibs committed:
1. Ditch currentPostcodeValid - only used for valid/invalid styling, which we can achieve with pure CSS
2. Shift constituencyIndex into the formData state rathe than having it's own separate state var
3. Add a useMemo version of lastSelectedConstituency (should duplicate/replace userConstituency)
4. Add a useRef version of currentPostcode (should duplicate/replace currentPostcode)

asibs committed:
Remove old currentPostcode & userConstituency vars

asibs committed:
Check in missing files

asibs committed:
Prettier fixes

asibs committed:
Merge branch 'refactor-lookup-component' into as/refactor-lookup-component-experiments

jms301 committed:
Fixed some edge cases and refactored a bit:

* If clearTimeout was called cancelling a pending fetch the promise never resolved & was in cache.
 - Cleared the cache at the same time as calling clearTimeout
 - Added a second setTimeout which checks if the 1st was cancelled and resolves the promise with null if it was.

* setError was not cleared when a new fetch was made so old errors could hang around.

* Removed lastValidPostcode (it wasn't necessary).
* renamed currentPostcode to validPostcode (currentPostcode.current felt unnecessary)

jms301 committed:
remove debug code & logging

asibs committed:
Merge branch 'main' into refactor-lookup-component
